### ISSUE TITLE
Build failure

### ISSUE THREAD

**Issue by jal-frezie** [APSIMInitiative/ApsimX#4772]
Just trying to build ApsimX on my new laptop (debian buster) -- followed all instructions in https://apsimnextgeneration.netlify.com/development/contribute/compile/linux/ and monodevelop gives 1 warning and 2 errors:

[Build Output 3:46 pm 2020-03-02.bin.log](https://github.com/APSIMInitiative/ApsimX/files/4276269/Build.Output.3.46.pm.2020-03-02.bin.log)

**Comment by hol430** [comment]
Hi Jasper - I'm currently abroad and don't have access to my external hdd with my debian vm. I'll look into this in a few weeks when I get back.

**Comment by zur003** [comment]
ApsimX mostly builds OK for me on Ubuntu 18.04. What format is your log file in? I can't make much sense of it.
I do get errors when trying to build UnitTests, where it appears to be trying to reference .Net assemblies for Gtk3 rather than Gtk2.

**Comment by jal-frezie** [comment]
The file I attached is the binlog output from monodevelop, changed to bin.log so I could upload it. It seems to be in gzip format, and I unzipped it after adding .gz extension. The gunzipped version has some legible ASCII. 
I think Debian Buster is about parallel with Ubuntu Bionic (18.04) in terms of package versions, but Debian has older versions of some things.

**Comment by jal-frezie** [comment]
Note that I cannot simply install the pre-built version because my distro does not include libwebkitgtk -- it only has libwebkit2gtk

**Comment by zur003** [comment]
Ouch. Not having libwebkitgtk is a big problem. I think libwebkit2gtk is based on Gtk 3, but ApsimNG is currently Gtk 2. How are you planning to circumvent this difficulty?

I had a little success at reading the binlog file, but it didn't offer many clues. It appears to be a log for an attempted build of ApsimNG, which failed because neither APSIM.Shared.dll nor Models.exe could be found. The projects need to be built in order: APSIM.Shared first, then Models, and finally ApsimNG. A log from an attempt to build just APSIM.Shared.dll might be more useful.

There was an indication in the log that APSIM.Shared might have failed to build because it couldn't find MathNet.Numerics. That should have been fetched by nuget, but something may have gone wrong. Try running `nuget restore` again, and take note of any messages indicating problems.

**Comment by jal-frezie** [comment]
OK, I built from the .csproj files in the order you suggested, and it all worked OK. Half the GUI seems to be missing, probably because I have no libwebkitgtk. I installed version 1.0 on my old system from an earlier .deb file but this is called a 'frankendebian' and frowned upon. The correct procedure is to build the missing library from source. I shall give that a go!

**Comment by hol430** [comment]
@zur003 what's the best way forward on this one? Compiling webkitgtk2 takes several hours for me (even with -j4 in makeflags), which is understandable but rather unfortunate. I'm unsure how big of a problem this will be on other distros - on arch I was able to install a prebuilt version which saved me some time. 

I guess in the long run if we make the move to gtk3 this will solve itself.

**Comment by zur003** [comment]
Yep. That's a major reason for thinking about moving to Gtk3 - Linux distros seem to be gradually dropping support for Gtk2. That's likely to only get worse.
The flip side, though, is that on Windows a lot of Gtk stuff is stuck at Gtk 2 (e.g., Mono on Windows), and much of that is 32-bit (I had to build the "glue" DLLs that we're using for 64-bit Windows Gtk 2 myself; I really should do a better job of documenting how that was done before I retire).
Using GtkSharp for the GUI gives us the advantages of C# (which I really like as a language and associated libraries) and cross-platform execution, but unfortunately it's not at all a mainstream approach, and the path forward is not nearly as clear as I'd like.

**Comment by jal-frezie** [comment]
Would someone be so kind as to summarize the problem here? I installed the pre-built version (2020.2.28.4761) on my Debian box, and it works very nicely -- the only non-standard software I have is the Mono project repositories. Does that version have the old webkitgtk statically linked or something?

**Comment by hol430** [comment]
How did you manage to install the prebuilt version? There should be a depends clause on libwebkitgtk which prevents dpkg from installing the package.

To summarise the problem - most distros are dropping support for libwebkitgtk, as it's quite old. Removing this dependency from apsim is going to be difficult. The interim solution is probably to manually install libwebkitgtk - either by compiling it or finding a prebuilt version.

**Comment by jal-frezie** [comment]
I added libwebkitgtk2.4.11-3 package and its dependencies from a different distribution (the 'Frankendebian' route). 
How badly does Apsim actually need any webkitgtk feature?

**Comment by zur003** [comment]
We need webkitgtk (or its equivalent) in several places where we are using HTML to format material for presentation to the user. The most obvious examples are in presenting the "summary file" and maps. It's also used to present the formatted Markdown in memos, and in several portions of the CLEM interface.
Using HTML is a very convenient way to lay out complicated documents, provided you have the means to render it. We could re-factor things to display this material in some other way, but it would require a fair bit of effort.

**Comment by jal-frezie** [comment]
OK, have pulled the current version and tried to build again -- build went OK, but when running the result I get this error:
System.TypeInitializationException: The type initializer for 'WebKit.WebView' threw an exception. ---> System.DllNotFoundException: webkit-1.0 assembly:<unknown assembly> type:<unknown type> member:(null)
  at (wrapper managed-to-native) WebKit.Download.webkit_download_get_type()
  at WebKit.Download.get_GType () [0x00000] in <bdb3830a765a4a11b18d7fd3d8a8ec71>:0 
  at GtkSharp.WebkitSharp.ObjectManager.Initialize () [0x00011] in <bdb3830a765a4a11b18d7fd3d8a8ec71>:0 
  at WebKit.WebView..cctor () [0x00000] in <bdb3830a765a4a11b18d7fd3d8a8ec71>:0 
   --- End of inner exception stack trace ---
  at (wrapper managed-to-native) System.Object.__icall_wrapper_mono_generic_class_init(intptr)
  at UserInterface.Views.TWWebBrowserWK.InitWebKit (Gtk.Box w) [0x00009] in <1d818aef1d974126b21c7942acd09d1e>:0 
  at UserInterface.Views.TWWebBrowserWK..ctor (Gtk.Box w) [0x0002c] in <1d818aef1d974126b21c7942acd09d1e>:0 
  at UserInterface.Views.HTMLView.CreateWebKitBrowser (Gtk.Box box) [0x00001] in <1d818aef1d974126b21c7942acd09d1e>:0 
  at UserInterface.Views.HTMLView.PopulateView (System.String contents, System.Boolean isURI) [0x0004e] in <1d818aef1d974126b21c7942acd09d1e>:0 
  at UserInterface.Views.HTMLView.SetContents (System.String contents, System.Boolean allowModification, System.Boolean isURI) [0x00026] in <1d818aef1d974126b21c7942acd09d1e>:0 
  at UserInterface.Presenters.GenericPresenter.Attach (System.Object model, System.Object view, UserInterface.Presenters.ExplorerPresenter explorerPresenter) [0x000ec] in <1d818aef1d974126b21c7942acd09d1e>:0 
  at UserInterface.Presenters.ExplorerPresenter.ShowInRightHandPanel (System.Object model, UserInterface.Views.ViewBase newView, UserInterface.Presenters.IPresenter presenter) [0x00041] in <1d818aef1d974126b21c7942acd09d1e>:0 
  at UserInterface.Presenters.ExplorerPresenter.ShowInRightHandPanel (System.Object model, System.String viewName, System.String presenterName) [0x00032] in <1d818aef1d974126b21c7942acd09d1e>:0 
  at UserInterface.Presenters.ExplorerPresenter.ShowRightHandPanel () [0x00184] in <1d818aef1d974126b21c7942acd09d1e>:0 
  at UserInterface.Presenters.ExplorerPresenter.OnNodeSelected (System.Object sender, UserInterface.Interfaces.NodeSelectedArgs e) [0x00008] in <1d818aef1d974126b21c7942acd09d1e>:0 
  at UserInterface.Views.TreeView.OnAfterSelect (System.Object sender, System.EventArgs e) [0x0006e] in <1d818aef1d974126b21c7942acd09d1e>:0 

...and I do not get the option to run a model. I have libwebkitgtk 1.0 installed (see above) and the latest .deb (5091) installs and runs OK, but I would of course like to compile it myself.

**Comment by hol430** [comment]
Can you post the output of running `MONO_LOG_LEVEL=debug mono Bin/ApsimNG.exe` up until the point where you get the error?

**Comment by zur003** [comment]
Just a hunch as to what may have gone wrong - you may need to copy webkit-sharp.dll.config from the ApsimNG/Assemblies folder into the folder that contains your ApsimNG.exe executable. You may also need to edit that file. The config file is used by the Mono runtime to locate native-mode shared libraries. It performs a mapping so that rather than looking for webkit-1.0.dll (appropriate only on WIndows), it looks for libwebkitgtk-1.0.so.0. You may need to modify this if the .so for libwebkitgtk has a different name on your system.
I'm just guessing that this is where the problem lies. Follow the suggestion of @hol430 to obtain log output which will give you information about exactly where the application is searching for the webkit shared library.

**Comment by jal-frezie** [comment]
Duh, my stupid fault -- I copied webkit-sharp.dll over, not 
webkit-sharp.dll.config. Did the latter and it now runs OK. Sorry guys!

**Comment by hol430** [comment]
We should add a pre-build step to do this automatically on Linux systems, similar to what we do on windows to copy the gtk-sharp binaries into the Bin directory. @zur003 do Mac users need to copy these webkit-sharp files into Bin/ as well?

**Comment by zur003** [comment]
Normally you shouldn't need to do this on Linux systems, as the relevant .config files are co-located with their managed assemblies in the Global Assembly Cache. Generally it's best to just let the gac do its thing, as it should "know" about local file names and paths. In Jasper's case, with a "bootleg" webkit-sharp.dll, there are a number of things that might have gone wrong. If webkit-sharp.dll has been properly installed in his gac, then he probably could just remove both webkit-sharp.dll and webkit-sharp.dll.config from his bin folder, and mono should pick everything up from the gac copy. 

On Macs, we're not using webkit-sharp - we're going through some Mac-specific stuff to embed Safari (which is webkit-based, but that's irrelevant). Mono on Mac _might_ need to see webkit-sharp.dll, since it's referenced, but should never actually need to invoke it - I really don't recall the details.

All the .config file stuff is Mono-specific, by the way. This will all need to be handled differently if and when we move to .Net Core 3.

**Comment by zur003** [comment]
@hol430 - Ah, sorry, I wasn't thinking about this correctly. You're talking about the case where the user is actually _building_ Apsim on Linux. It might well be possible to do the copying of webkit-sharp.dll, Mono.TextEditor.dll, and their associated config files automatically in that situation.

**Comment by hol430** [comment]
@zur003 are we supposed to be pulling in the *-sharp.dll assemblies inside DeploymentSupport/Windows/ when compiling on Linux? I ask because here is an excerpt from a recent build of ApsimNG on my machine:

```
Primary reference "atk-sharp, Version=2.12.0.0, Culture=neutral, PublicKeyToken=35e10195dab3c99f". (TaskId:107)
      Resolved file path is "/home/drew/code/ApsimX2/DeploymentSupport/Windows/Bin64/atk-sharp.dll". (TaskId:107)
      Reference found at search path location "{HintPathFromItem}". (TaskId:107)
      This reference is not "CopyLocal" because at least one source item had "Private" set to "false" and no source items had "Private" set to "true". (TaskId:107)
      The ImageRuntimeVersion for this reference is "". (TaskId:107)
```

I have installed gtk-sharp-2 2.12.45-2 installed.

Worth noting that the HintPaths used in ApsimNG points to DeploymentSupport, whereas the HintPaths in the UnitTests project points to `../../Bin/` which causes the unit tests to fail to compile on my machine because it doesn't search in DeploymentSupport. In fact, the UnitTests project doesn't find gtk-sharp2 in the GAC - it finds gtk-sharp3; I can fix this by changing the HintPath in UnitTests.csproj to point towards DeploymentSupport but it begs the question of why this is necessary.

```
Primary reference "atk-sharp, Version=3.0.0.0, Culture=neutral, PublicKeyToken=35e10195dab3c99f". (TaskId:154)
      Resolved file path is "/usr/lib/mono/gac/atk-sharp/3.0.0.0__35e10195dab3c99f/atk-sharp.dll". (TaskId:154)
      Reference found at search path location "{GAC}". (TaskId:154)
          For SearchPath "{CandidateAssemblyFiles}". (TaskId:154)
          Considered "/home/drew/code/ApsimX2/packages/NUnit3TestAdapter.3.9.0/build/net35/NUnit3.TestAdapter.dll",
  			but its name "NUnit3.TestAdapter"
  			didn't match the expected name "atk-sharp, Version=3.0.0.0, Culture=neutral, PublicKeyToken=35e10195dab3c99f". (TaskId:154)
          Considered "/home/drew/code/ApsimX2/packages/NUnit3TestAdapter.3.9.0/build/net35/nunit.engine.dll",
  			but its name "nunit.engine"
  			didn't match the expected name "atk-sharp, Version=3.0.0.0, Culture=neutral, PublicKeyToken=35e10195dab3c99f". (TaskId:154)
          Considered "/home/drew/code/ApsimX2/packages/NUnit3TestAdapter.3.9.0/build/net35/nunit.engine.api.dll",
  			but its name "nunit.engine.api"
  			didn't match the expected name "atk-sharp, Version=3.0.0.0, Culture=neutral, PublicKeyToken=35e10195dab3c99f". (TaskId:154)
          Considered "/home/drew/code/ApsimX2/packages/NUnit3TestAdapter.3.9.0/build/net35/Mono.Cecil.dll",
  			but its name "Mono.Cecil"
  			didn't match the expected name "atk-sharp, Version=3.0.0.0, Culture=neutral, PublicKeyToken=35e10195dab3c99f". (TaskId:154)
          Considered "/home/drew/code/ApsimX2/packages/NUnit3TestAdapter.3.9.0/build/net35/Mono.Cecil.Pdb.dll",
  			but its name "Mono.Cecil.Pdb"
  			didn't match the expected name "atk-sharp, Version=3.0.0.0, Culture=neutral, PublicKeyToken=35e10195dab3c99f". (TaskId:154)
          Considered "/home/drew/code/ApsimX2/packages/NUnit3TestAdapter.3.9.0/build/net35/Mono.Cecil.Mdb.dll",
  			but its name "Mono.Cecil.Mdb"
  			didn't match the expected name "atk-sharp, Version=3.0.0.0, Culture=neutral, PublicKeyToken=35e10195dab3c99f". (TaskId:154)
          Considered "/home/drew/code/ApsimX2/packages/NUnit3TestAdapter.3.9.0/build/net35/Mono.Cecil.Rocks.dll",
  			but its name "Mono.Cecil.Rocks"
  			didn't match the expected name "atk-sharp, Version=3.0.0.0, Culture=neutral, PublicKeyToken=35e10195dab3c99f". (TaskId:154)
          For SearchPath "{HintPathFromItem}". (TaskId:154)
          Considered "/home/drew/code/ApsimX2/Bin/atk-sharp.dll", but it didn't exist. (TaskId:154)
          For SearchPath "{TargetFrameworkDirectory}". (TaskId:154)
          Considered "/usr/lib/mono/xbuild-frameworks/.NETFramework/v4.6/atk-sharp.winmd", but it didn't exist. (TaskId:154)
          Considered "/usr/lib/mono/xbuild-frameworks/.NETFramework/v4.6/atk-sharp.dll", but it didn't exist. (TaskId:154)
          Considered "/usr/lib/mono/xbuild-frameworks/.NETFramework/v4.6/atk-sharp.exe", but it didn't exist. (TaskId:154)
          Considered "/usr/lib/mono/4.6-api/atk-sharp.winmd", but it didn't exist. (TaskId:154)
          Considered "/usr/lib/mono/4.6-api/atk-sharp.dll", but it didn't exist. (TaskId:154)
          Considered "/usr/lib/mono/4.6-api/atk-sharp.exe", but it didn't exist. (TaskId:154)
          Considered "/usr/lib/mono/4.6-api/Facades/atk-sharp.winmd", but it didn't exist. (TaskId:154)
          Considered "/usr/lib/mono/4.6-api/Facades/atk-sharp.dll", but it didn't exist. (TaskId:154)
          Considered "/usr/lib/mono/4.6-api/Facades/atk-sharp.exe", but it didn't exist. (TaskId:154)
          Considered "/usr/lib/mono/4.6-api/Facades/atk-sharp.winmd", but it didn't exist. (TaskId:154)
          Considered "/usr/lib/mono/4.6-api/Facades/atk-sharp.dll", but it didn't exist. (TaskId:154)
          Considered "/usr/lib/mono/4.6-api/Facades/atk-sharp.exe", but it didn't exist. (TaskId:154)
      Found related file "/usr/lib/mono/gac/atk-sharp/3.0.0.0__35e10195dab3c99f/atk-sharp.dll.config". (TaskId:154)
      This reference is not "CopyLocal" because it's registered in the GAC. (TaskId:154)
      The ImageRuntimeVersion for this reference is "". (TaskId:154)
```

**Comment by zur003** [comment]
@hol430 - The cross-platform world can get a bit messy. Normally an install of gtk-sharp2 on Linux will install atk-sharp.dll (and its partners glib-sharp, gdk-sharp, gtk-sharp, pango-sharp and Mono.Cairo) into your GAC. The installer also includes the appropriate .config files for each, to enable them to find the native .so files they require. On Windows, though, we don't make the assumption that these files are in the GAC; we copy all the necessary DLLs (both managed and native) into the Bin folder. 
We need to think about the situation at compile time vs. the situation at run time. On WIndows at run-time, all the binaries should be in the Bin folder. On Linux at run time, if properly installed, the Gtk support will be in the GAC. If not installed, then the user would need to manually copy in the managed Gtk dlls, and add .config files for each.
What about at compile time? On Windows, for the user interface we point to the copies sitting in DeploymentSupport\Windows\Bin64 as references. That works on Linux, too. I'm not sure why the UnitTests are set up differently, unless it is because we normally run the unit tests on Windows, and pointing to the Bin folder guarantees that we're using the same DLLs as will be used in the distribution. But note also that in UnitTests, CopyLocal is set, so it actually copies the various DLLs. This may not play well on Linux, even if the DLLs are in the Bin folder, as it won't also be copying the .config files.
So, in short, UnitTests was set up for use on Windows, and hasn't really been configured to play nicely on Linux (or OS X, for that matter). Perhaps it should be configured differently.
